src/predicate_lang/predicate_lang.ml

```diff
@@ -77,65 +77,111 @@ and many f k acc =
 
 and decode f = many f union []
 
-let rec encode f =
+let rec encode f t =
   let open Encoder in
-  function
+  match t with
```

Comment:
There are still some left-over diffs here that would be nice to clean up.

Reply:
done